### PR TITLE
restructure elasticsearch process management

### PR DESCRIPTION
**Pull request by thrau**
This PR completely restructures the startup procedure for the elasticsearch service. It introduces a object-oriented-style wrapper around the elasticsearch binary and the involved installation procedure. This makes it reusable, simplifies the use inside `es_api.py`, and makes it easier in the future to add multi-cluster implementation (one cluster per domain). I also added proper concurrency control that was previously missing. I kept the principle that there is only one cluster for all domains, but made it more explicit in the code, and added comments indicating where changes have to be made, if we one day want a setup where each domain can spawn its own cluster (for that we would have to also refactor the installation procedure, as we assume a single installation + data directory).

I haven't tested it with the `DATA_DIR` option yet

this is a useful code snippet to test manually:

```python
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)

    es_client = aws_stack.connect_to_service("es")
    name = 'my domain name'

    ## run afterwards
    # es_client.delete_elasticsearch_domain(DomainName=name)
    # exit(0)

    # status = es_client.describe_elasticsearch_domain(DomainName=name)
    # LOG.info('domain status %s', status)

    kwargs = {}
    LOG.info('creating elasticsearch domain %s', name)
    es_client.create_elasticsearch_domain(DomainName=name, **kwargs)
    print(es_client.list_domain_names()["DomainNames"])

    # wait for completion status
    def check_cluster_ready(*args):
        LOG.info('describe_elasticsearch_domain %s ', name)
        status = es_client.describe_elasticsearch_domain(DomainName=name)
        created = status["DomainStatus"]["Created"]
        LOG.info('asserting created state of domain %s (state = %s)', name, created)
        assert created

    LOG.info('waiting for domain to be ready...')
    retry(check_cluster_ready, sleep=10, retries=12)
    LOG.info('created!')
```

### CODE REVIEW

**Comment by mgagliardo** [comment]
I really  like this approach very much but I do have two concerns only:

1. Would not this be 100% backward incompatible with the option of using your own `ELASTICSEARCH_BACKEND`? Not all cases but some of them there is a need of setting up your own ES node for multi-node and/or further testing as what we currently provide (and still do) is a default single-noded cluster.

2. Not sure about this one so I would prefer to ask, is there still a chance to install ES plugins to this cluster or does this modify that default process as well? 

All in all it sounds great!

**Comment by thrau** [comment]
thanks for the feedback! should have requested you as reviewer in the first place :-) 

i'm not sure how `ELASTICSEARCH_BACKEND` works. i grepped the codebase but couldn't find a reference to it. where is it currently used? but if i understand it correctly, when `ELASTICSEARCH_BACKEND` is provided, localstack should not start elasticsearch, but rather just point the proxy to the `ELASTICSEARCH_BACKEND`? if that's the case then we only need to modify this bit to not start a cluster, but simply use the config of the provided `ELASTICSEARCH_BACKEND`: https://github.com/localstack/localstack/blob/1b1374c824a45ddbed8bf68f4d9b256d15a0e9a1/localstack/services/es/cluster.py#L318-L321

not sure about the plugin install. i didn't change anything in the install routine. it's still using `localstack.services.install.install_elasticsearch` to install elasticsearch.

File: tests/integration/test_elasticsearch.py [hunk 1]
```diff
@@ -52,6 +55,11 @@ def tearDownClass(cls):
             d["DomainName"] for d in es_client.list_domain_names()["DomainNames"]
         ]
 
+    def test_create_existing_domain_causes_exception(self):
+        # already created in
```
Comment (whummer) [inline tests/integration/test_elasticsearch.py:59]: nit: could be changed to
```
# already created in setUpClass()
```

**Review by whummer — APPROVED** [review]
Looks great, thanks for this refactoring @thrau ! 👍

File: localstack/services/es/cluster.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+            # FIXME: if this fails the cluster could be left in a wonky state
+            # FIXME: this is not a good place to run install, and it only works because we're
+            #  assuming that there will only ever be one running Elasticsearch cluster
+            install.install_elasticsearch(self.version)
+            self._init_directories()
+
+            cmd = self._create_run_command(additional_settings=self.command_settings)
+            cmd = " ".join(cmd)
+            if is_root():
+                # FIXME: this is a very strong assumption
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/es/cluster.py:151]: i like that! will add

File: localstack/utils/common.py [hunk 3]
```diff
@@ -720,6 +720,29 @@ def retry(function, retries=3, sleep=1, sleep_before=0, **kwargs):
     raise raise_error
 
 
+def poll(condition, timeout: float = None, interval: float = 0.5) -> bool:
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/utils/common.py:723]: happy to reconcile them at some point. i think one could be implemented with the other, but i think it's useful to have the two separate. i renamed the method into `poll_condition` to make it a bit more explicit.

File: localstack/services/es/cluster.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+            self._elasticsearch_thread.stop()
+
+    def start(self):
+        with self._lifecycle_lock:
+            if self._started:
+                return
+            self._started = True
+
+        threading.Thread(target=self._run_elasticsearch, daemon=True).start()
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/es/cluster.py]: not a huge fan of the abstraction and the complexity/magic it adds. but replaced it nonetheless :+1:

[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/es/cluster.py]: not a huge fan of the abstraction and the complexity/magic it adds. but replaced it nonetheless :+1:

**Comment by thrau** [comment]
@mgagliardo I'm happy to make follow up changes to accommodate the features if you let me know how exactly it worked previously

**Pull request merged**